packages/deployer/internal/ast/router-validator.js

```diff
@@ -0,0 +1,113 @@
+const fs = require('fs');
+const path = require('path');
+const { getCaseSelectors } = require('./ast-helper');
+const relativePath = require('@synthetixio/core-js/utils/relative-path');
+const filterValues = require('filter-values');
+const { getSelectors } = require('@synthetixio/core-js/utils/contracts');
+const logger = require('@synthetixio/core-js/utils/logger');
+
+async function getContractSelectors(onlyModules) {
+  const modules = onlyModules
+    ? filterValues(hre.deployer.deployment.general.contracts, (c) => c.isModule)
+    : hre.deployer.deployment.general.contracts;
+  const modulesNames = Object.keys(modules);
+
+  const items = [];
+  for (const moduleName of modulesNames) {
+    const moduleArtifacts = await hre.artifacts.readArtifact(moduleName);
+    const selectors = await getSelectors(moduleArtifacts.abi);
+    selectors.forEach((selector) => {
+      items.push({ selector, moduleName });
+    });
+  }
+  return items;
+}
+
+function getRouterSource() {
+  const routerPath = path.join(
+    relativePath(hre.config.paths.sources, hre.config.paths.root),
+    'Router.sol'
+  );
+
+  const source = fs.readFileSync(routerPath).toString();
+  return source;
+}
+
+async function findMissingSelectors() {
```

Comment:
Might be clearer to rename to `findMissingSelectorsInSource()`. Maybe there's a better name. My point is to hint that this function will be looking at source code instead of AST.

Also, you may want to consider moving this out of the AST folder.